Verlet rope improvements

lolleko opened the pull request:
Fixes #23 

- Probably still needs more iteration in the future. But ropes are now a lot less likely to clip.
- Also added an unstuck logic for ropes that are stretched, for cases where a verlet point gets stuck behind a wall or under ground.
- Performance should be about the same, probably even faster because the heavy lifting now runs in FixedUpdate rather than every frame.

Now:


https://github.com/user-attachments/assets/b31b7a3e-93b1-4f86-ae1f-963861be0591


Before:

https://github.com/user-attachments/assets/24873f85-b552-44b3-83df-356cc2e8cbab

Sl4vP0weR commented:
Not renamed folders.

lolleko committed:
Move simulation to fixedupdate

lolleko committed:
Split simulation into multiple steps & Accumulate point acceleration

lolleko committed:
Stiffness constraint, Improved collision check

lolleko committed:
Disable collision for stretched rope segments, to allow them to unstuck/recover

lolleko committed:
Additional checks to prevent ropes from clipping into the ground

garrynewman commented:
If this is in fixedupdate now, does that mean it's running at 30 fps? If fixed update is set at 5fps doesn't that look terrible?

<img width="674" height="460" alt="image" src="https://github.com/user-attachments/assets/f0d0faf1-db9a-4335-830b-6232b1a1610f" />

lolleko committed:
Add substeps to simulation, increase numeric stability

lolleko committed:
Send rope simulation to sleep if rope is not moving.

lolleko committed:
Run rope simulation with physics step

lolleko commented:
Fixed update rate is nice because it improves numeric stability.

But you are right we probably want to make sure it runs often enough.
I changed it so the simulation runs with every physics step, which has a frequency of 50 by default irrc.
Pretty sure this should also get rid of the artifact in your screenshot.
We can also add substepping to the rope simulation, but when I tried I didn't see any improvements.


Also added a resting logic so ropes that don't move don't consume any resources.
For optimization, long term we should probably move the simulation to a gameobjectsystem so it's easer to parallelize and optimize for CPU cache.

garrynewman commented:
Fixed update is fine - it just needs to introduce some kind of lerping. My monitor is 240hz, I don't want 50hz moving ropes.

lolleko committed:
Lerp verlet rope points, mimics physics interpolation

lolleko committed:
Apply friction to ropes that are sliding along a surface

lolleko committed:
Merge branch 'main' into verlet-rope-improvements

lolleko commented:
I should get one of those.
Totally forgot about lerping. Added some basic lerping, roughly mimicking what we are doing with our interpolated transforms. 

https://github.com/user-attachments/assets/d992c93f-9e09-407a-b334-a2b5a1881b3c



Also added friction to the ropes.

garrynewman commented:
Seems to get caught up way too much, this stuff shouldn't happen where the physical joint is at the point of pulling but the rope is wrapped around stuff, not reflecting what is happening at all.

https://github.com/user-attachments/assets/53d42b0f-58fb-4851-92c1-0829daf5a372

I get that this is a bit of a balance, but we have to lean towards it representing the joint wherever possible

lolleko committed:
Lower collision ignore thresholds

lolleko commented:
It really is tricky because we want ropes to respect collisions and not clip into the ground, but at the same time if the rope is too stretched we want it to ignore collisions so it gets unstuck.

This behavior can be fine tuned using `collisionMaxRopeSegmentStretchFactor`, `collisionMaxRopeStretchFactor`. I lowered their values a bit, what you showed is now less likely to happen.
Feel free to play around with them yourself. Just keep in mind if set too low we might get some ground clipping.

garrynewman commented:
There's another problem, once the rope is reasonably long it gets way too saggy. These ropes should all have 0 slack on them.

![sbox 2025 07 25 12 34 30](https://github.com/user-attachments/assets/03076aee-4124-466a-b4e0-d306d49d9862)

lolleko committed:
Tweak some values optimize constraint loop

lolleko committed:
Bidirectional constraint solving for faster convergence

lolleko committed:
Improve unstuck logic

lolleko committed:
Adaptive Constraint Iterations

lolleko committed:
Improved surface sliding and friction for ropes

lolleko committed:
Lower stiffnes

lolleko committed:
Fixed compute/segment budget for every rope

lolleko committed:
Scale rest thresholds based on segment length

lolleko committed:
Tweak some more values

lolleko committed:
Use GameobjectSystem to run rope simulation in parallel

https://files.facepunch.com/lolleko/2025/July/29_21-19-GlamorousHylaeosaurus.mp4

lolleko committed:
Optimize rope sim to fit in a ~1ms budget.

https://files.facepunch.com/lolleko/2025/July/30_10-44-VioletredTeal.mp4

lolleko committed:
Refactor, docs, final value tweak

lolleko commented:
I am pretty happy with this now.

The rope getting caught up on stuff is fixed, the ground clipping is fixed and the ropes have 0 slack when stretched.
Maybe give it another try and see if you can notice any more issues.

https://github.com/user-attachments/assets/2768d0d8-3ee6-4cbe-aac1-975b1cc66e38

https://github.com/user-attachments/assets/712bbab5-1112-4662-87c0-0965e8c465c0


I optimized the simulation to fit into a ~0.5ms budget for ~50 (active/nonsleeping) ropes on my CPU (Ryzen 9 7900X).
There are some ways to bring this further down, if it proves to be an issue on lower end hardware.
But those would come at the cost of visual quality and/or code complexity, so I would only do those if they are actually needed.

https://github.com/user-attachments/assets/23c62625-a2a6-4b08-90c0-e91af6d965c5

lolleko committed:
Rename Width to Radius

lolleko committed:
Reorder functions in class to match execute order in simulate

garrynewman commented:
(merged)
